### ISSUE TITLE
Encoded values do not match after adding TagParser

### ISSUE THREAD

**Issue by Coneys** [graphhopper/graphhopper#2617]
**Describe the bug**
After upgrading graphhopper to 5.3, adding TagParser to encodingManagerBuilder makes different configuration on first run, and on next runs, which resolves in "Encoded values do not match" error.

Difference is in shift value (and version)
Graphhopper config: surface|version=782288784|bits=4|min_value=0|negate_reverse_directionfalse|index=0|shift=4|store_both_directions=false
Graph: surface|version=778378475|bits=4|min_value=0|negate_reverse_directionfalse|index=0|shift=0|store_both_directions=false

Before everything worked, because I could create EncodingManager from code. Now there is no way of doing this. 
Also after hours of debugging I saw that in first run only one EncodedValue is created for Surface (from TagParser). In next run there are two, from TagParser AND from DefaultEncodedValueFactory 

**To Reproduce**

Create Graphhopper from version 5.3 using Java code with OSMSurfaceParser:

```
 return GraphHopper().apply {
            
            ...
            
            val surfaceParser = OSMSurfaceParser()
            
            encodingManagerBuilder.add(surfaceParser)

            importOrLoad()
        }
```

For the first run, everything works and I get surface on results. In the next load, error occures

**Expected behavior**

There should be no error, like in the previous versions

**System Information**

GraphHopper version 5.3

**Screenshots & Logs**

To create Graphhopper instance, I am using custom EncoderFactory, Elevation provider and profiles, but it should not be relevant.

Thanks for help, and thanks for that great library ;) Hope we can resolve this issue

**Comment by easbar** [comment]
The first thing I would try is calling importOrLoad, instead of load, on your second run. I'll try to reproduce this error, is the code you shared really all I need?

**Comment by easbar** [comment]
> I didn't know that I should use "load" on second run.

Yes, I think you should not except in a few special cases where you have to unfortunately. Can you share some runnable code that reproduces your error?

**Comment by Coneys** [comment]
Oh I see. Yes I was using importOrLoad always 
Code below is enough to make it crash on second run:

` fun issueForGithub(graphDirectory: File, osmDataFile: File): GraphHopper {
        return GraphHopper().apply {
            osmFile = osmDataFile.absolutePath
            graphHopperLocation = graphDirectory.absolutePath
            profiles = listOf(Profile("car"))
            val surfaceParser = OSMSurfaceParser()
            encodingManagerBuilder.add(surfaceParser)
            importOrLoad()
        }
    }`

only dependency I am using is this:

implementation("com.graphhopper:graphhopper-core:5.3")

**Comment by Coneys** [comment]
2.3

**Comment by easbar** [comment]
Ok just for the record here is the code that reproduces the error in Java:

```java
    @Test
    void issue2617() {
        {
            GraphHopper hopper = new GraphHopper();
            hopper.setOSMFile(BAYREUTH)
                    .setGraphHopperLocation(GH_LOCATION)
                    .setProfiles(new Profile("car"));
            hopper.getEncodingManagerBuilder()
                    .add(new OSMSurfaceParser());
            hopper.importOrLoad();
        }
        System.out.println("load");
        {
            GraphHopper hopper = new GraphHopper();
            hopper.setOSMFile(BAYREUTH)
                    .setGraphHopperLocation(GH_LOCATION)
                    .setProfiles(new Profile("car"));
            hopper.getEncodingManagerBuilder()
                    .add(new OSMSurfaceParser());
            hopper.importOrLoad();
        }
    }
```

`hopper.getEncodingManagerBuilder()` was added in #2281 so it did not exist in GH 2.x and was added for GH 3.0. The way you use the method causes the surface encoded value to be added twice to the encoding manager (and it does not throw an error). Basically this does not throw an error, but the encoded value strings will be different (I used max_width instead of surface but that does not matter):

```java
    @Test
    public void issue2617() {
        EncodingManager.Builder emBuilder = EncodingManager.start();
        emBuilder.add(new OSMMaxWidthParser());
        EncodingManager em = emBuilder.build();
        String emString1 = em.toEncodedValuesAsString();

        emBuilder = EncodingManager.start();
        emBuilder.add(new OSMMaxWidthParser());
        // from graph.encoded_values
        emBuilder.add(MaxWidth.create());
        em = emBuilder.build();
        String emString2 = em.toEncodedValuesAsString();

        assertEquals(emString1, emString2);
    }
```

In the first run only the parser is added to the encoding manager builder, but in the second run first the encoded values are added (taken from graph.encoded_values in the properties file) and then because of the added parser the surface encoded value is added again. This bug already exists since #2281. Anyway, in current master `hopper.getEncodingManagerBuilder()` was already removed.

So your workaround of not adding the parser for the second run is indeed a good idea. Maybe a safer way is using the init() method before calling `importOrLoad`. This way you avoid the code path that involves the `EncodedValueFactory` (also see #2438  if you are really interested), because the encoding manager will already be built in the `init()` method. For example you can add the surface encoded value like this (use the same for both runs):

```java
hopper.init(new GraphHopperConfig().putObject("graph.encoded_values", "surface,max_width,..."));
hopper.importOrLoad();
```

But be careful when you combine the `init` method with the different setter methods of the `GraphHopper` class. When in doubt call the init method first so it does not overwrite the results of the setter method calls, see also this comment: c79bb964b42fb9416fb719ec22c1c6ae575b7a39

**Comment by Coneys** [comment]
Thanks for your response. I prefer not to use init() because of no type safety properties, but I understand issue and if adding TagParser only once is good idea, then I'll stick with it.

Could you provide some example how should it be done in new version (which you refer as current master)?
I could not find any "fresh" example of how to add new TagParser or modify EncodingManager. 
Maybe new method in [RoutingExample.java](https://github.com/graphhopper/graphhopper/blob/master/example/src/main/java/com/graphhopper/example/RoutingExample.java)?

**Comment by easbar** [comment]
You should be able to find examples for most things in `GraphHopperTest.java` or `GraphHopperOSMTest.java`, for you specific version. 

> should it be done in new version (which you refer as current master)?

The current master is working most of the time, but can be subject to frequent changes and currently everything regarding the EncodingManager is moving quite fast, so I'd still recommend to use version 5.3 in most cases. Version 6.x will be released around September.

For current master the way to modify the EncodingManager ist to use `setFlagEncoderFactory`, `setEncodedValueFactory` and `setTagParserFactory` to define what kind of FlagEncoders, EncodedValues and TagParsers shall be *available* and then use `setFlagEncodersString` and `setEncodedValuesString` to define which ones should be *activated*. But also note that the meaning of the EncodingManager changed. It used to include the tag parsers, but now it only include encoded values and everything related to OSM parsing is in a new class called OSMParsers, which is not needed at all for the second run. Also the EncodingManager is loaded from disk in the second run, so there is no longer a need to make sure it is the same in the first and second run.

If all you want to do is adding the surface encoded value you do not need to do anything at all, because it is always added by default.

**Comment by Coneys** [comment]
Sounds great! In current version it was really difficult to understand which code should I use to tell what is available, and which configures what should be created. 
About "If all you want to do is adding the surface encoded value you do not need to do anything at all, because it is always added by default", you are talking about new version which will be out around September? Because on 5.3 surface is not encoded by default

**Comment by easbar** [comment]
> Sounds great! In current version it was really difficult to understand which code should I use to tell what is available, and which configures what should be created.

Yes, documentation etc. still needs to be updated for GH 6 as there have been many changes and even more might follow. Maybe the GraphHopper class will be removed entirely.

> Because on 5.3 surface is not encoded by default

Yes, you are right, the surface encoded value is not in the list of default encoded values. Neither for 5.3 nor for master: https://github.com/graphhopper/graphhopper/blob/af5ac0b0ae024da4b23966b1eef480508b5eedbb/core/src/main/java/com/graphhopper/routing/util/EncodingManager.java#L271-L284

**Comment by easbar** [comment]
> which code should I use to tell what is available, and which configures what should be created.

Like I said, what is available is defined in FlagEncoder/EncodedValue/TagParserFactory and what is created in flagEncoders/encodedValuesString. Except the fact that the EncodingManager.Builder adds some default encoded values as well, which I think it should not. Does this not make sense, or which version did you think was difficult to understand?

**Comment by easbar** [comment]
> Anyway, after reading your description of version 6.x, it doesn't matter, because it will be changed anyway ;)

It still matters, because feedback like yours helps improving the project, thanks!
